### PR TITLE
[READY] Mafia, a minigame for ghosts to sign up for while dead

### PR DESCRIPTION
**Pull request by tralezab**
### thanks to anturk for the base of all of this, i just polished things up for the ball and again for master

it's town of salem, it's werewolf, it's mafia. you got people in a funny little town trying to find out who is killing people every night using special town powers.

uh... what else... i moved the joining system off of the signups and onto a ghost button to join the game, and when enough people are in it starts. people who disconnect or rejoin the game while signed up are removed from the list when it thinks there are enough players (but not enough valid players)

## For the future:
Special hud for mafia players instead of tgui panels (though those will exist for role lookups)
And... this list of things I plan here, sorry it's kinda raw
image

## Last thing:
Reason why I put it up earlier than I'd normally like to is because any reviews to improve the code off of rushed-summer-ball-hoohah is super appreciated

## Why It's Good For The Game

approved by oranges, mafia is a really fun and based minigame, and now we have something more than just CTF. lot of potential to work off of here and anturk's amazing work making this possible should be realized

## Changelog
:cl: Anturk + Arm + other summer ball contributors code, trollbreeder + fikou + jerryderpington mapping
add: Mafia minigame added for ghosts
/:cl:

### CODE REVIEW

File: code/modules/mafia/_defines.dm [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	name = "9 Player Setup C"
+	roles = list(
+		/datum/mafia_role=3,
+		/datum/mafia_role/warden=1,
+		/datum/mafia_role/detective=1,
+		/datum/mafia_role/fugitive=1,
+		/datum/mafia_role/mafia=2
+	)
+
+/datum/mafia_setup/two_test
```
Comment (tralezab) [inline code/modules/mafia/_defines.dm:196]: this cannot exist when merged

File: code/modules/mafia/_defines.dm [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+	name = "12 Player Setup Lockdown"
+	roles = list(
+		/datum/mafia_role=4,
+		/datum/mafia_role/chaplain=1,
+		/datum/mafia_role/warden=2,
+		/datum/mafia_role/mafia=3
+	)
+
+/datum/mafia_setup/twelve_rip
+	name = "12 Player Setup rip"
```
Comment (ATH1909) [inline code/modules/mafia/_defines.dm:102]:
```suggestion
	name = "12 Player Setup RIP"
```

File: code/modules/mafia/controller.dm [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+	var/judgement_lynch_period = 5 SECONDS //just watch a guy die (or live but lol)
+	var/night_phase_period = 45 SECONDS //mafia talk and pick someone to kill
+	var/victory_lap_period = 20 SECONDS //just waiting around sayin' woop woop
+
+	var/list/current_setup_text //Redable list of roles in current game
+
+	var/game_id //Used to sync all parts - signup boards, spawns, night curtains, defaults to "mafia"
+	var/list/signed_up = list()
+
+	var/player_outfit = /datum/outfit/mafia //todo some fluffy outfit
```
Comment (ATH1909) [inline code/modules/mafia/controller.dm:22]: 👀

Comment (tralezab) [reply code/modules/mafia/controller.dm:22]: i kinda like the grey generic outfits people get. i might just keep it.

File: code/modules/mafia/_defines.dm (hunk 2 above)
Comment (tralezab) [reply code/modules/mafia/_defines.dm:102]: sure

File: _maps/map_files/Mafia/mafia_ball.dmm [hunk 4]
```diff
@@ -0,0 +1,48 @@
+"a" = (/turf/closed/indestructible/riveted,/area/mafia)
+"b" = (/turf/closed/indestructible/riveted/plastinum,/area/mafia)
+"c" = (/turf/closed/wall/rust,/area/mafia)
+"d" = (/turf/open/floor/plating,/area/mafia)
+"e" = (/turf/open/floor/plating{icon_state = "platingdmg3"},/area/mafia)
+"f" = (/turf/open/floor/plating{icon_state = "panelscorched"},/area/mafia)
+"g" = (/obj/mafia_game_board,/turf/open/floor/plating{icon_state = "platingdmg3"},/area/mafia)
+"h" = (/obj/structure/grille/indestructable,/turf/open/floor/plating,/area/mafia)
+"i" = (/turf/open/floor/plating{icon_state = "platingdmg1"},/area/mafia)
+"j" = (/obj/machinery/door/airlock/maintenance_hatch{max_integrity = 99999; name = "Maintenance"},/obj/effect/mapping_helpers/airlock/locked,/turf/open/floor/plating,/area/mafia)
```
Comment (ATH1909) [inline _maps/map_files/Mafia/mafia_ball.dmm:10]: use INFINITY instead of 99999, just in case

Comment (tralezab) [reply _maps/map_files/Mafia/mafia_ball.dmm:10]: good catch, thank you. i already caught one case of this earlier

**Comment by tralezab** [comment]
just gotta mapmerge and this will pass all checks - will do tomorrow though it's late

**Comment by ArcaneMusic** [comment]
Look at Mr. Tralezab over here, trying to re-add gangs yet again 😏

**Comment by Doosteroni** [comment]
I fukken loved town of Salem I’m lookin out for this shit

File: _maps/map_files/Mafia/mafia_ball.dmm (hunk 4 above)
Comment (nemvar) [reply _maps/map_files/Mafia/mafia_ball.dmm:10]: Map isn't map merged.

File: code/modules/mafia/_defines.dm [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+		/datum/mafia_role/traitor=3
+	)
+
+// 10 Player
+/datum/mafia_setup/ten_A
+	name = "10 Player Setup A"
+	roles = list(
+		/datum/mafia_role=6,
+		/datum/mafia_role/detective=1,
+		/datum/mafia_role/clown=1,/datum/mafia_role/mafia=3
```
Comment (MrPerson) [inline code/modules/mafia/_defines.dm:141]: Woops

Comment (tralezab) [reply code/modules/mafia/_defines.dm:141]: 😔 🤙

File: code/modules/mafia/controller.dm [hunk 6]
```diff
@@ -0,0 +1,528 @@
+/datum/mafia_controller
+	var/list/all_roles = list()
```
Comment (bobbahbrown) [inline code/modules/mafia/controller.dm:2]: needs dmdoc all over

Comment (tralezab) [reply code/modules/mafia/controller.dm:2]: 👍

**Comment by tralezab** [comment]
let's see if this compiles for the first test 🤓

**Comment by tralezab** [comment]
cmon stop conflciting you big jerk

**Comment by Jared-Fogle** [comment]
image
Don't agree with this button being here, would rather it's a vote system like CTF that pings everyone once enough votes has been reached.

**Comment by tralezab** [comment]
The mafia button does work like a vote system though, it only starts if enough people are joining it. I do really appreciate feedback though thank you

**Comment by Jared-Fogle** [comment]
I meant that I don't agree with the button being down there at all times, it's a bit obtrusive compared to CTF.

**Comment by tralezab** [comment]
Would you be alright with it if I made it a general minigame signup button and made CTF work the same way?

**Comment by Jared-Fogle** [comment]
I'd be fine with that since it'd at least be consistent, and would make way for more cool minigames (that I might even want to contribute).

**Comment by tralezab** [comment]
alright i'll go in that direction then, thanks. the next few testmerges will have the mafia only button until i change that to include CTF

**Comment by Jared-Fogle** [comment]
Bug report--The mafia zone bumped the station to a different z-level somehow, and it took over what should be the station z-level. Cult runes didn't work and a blob spawned in the mafia room instead.

dreamseeker_2020-06-24T16-01-56

**Comment by tralezab** [comment]
Fixed in next commit + a couple other things which make the game better for testing:
smaller chat
better signups?
dangerous species like plasmamen now switch to human
etc things i find

**Comment by CitrusGender** [comment]
spectacular work arm, glad to see you're getting this off the ground.

**Comment by TheVekter** [comment]
Mmmm, Mafia. Kill the detective night two. Convince the town you're the detective and start feeding them bad info. Be incredibly obvious as the Jester so everyone thinks you're faking it and kills you anyway, then you win.

**Comment by Jared-Fogle** [comment]
image
Is not preserving species intentional? ~~Because I look hideous~~

File: tgui/packages/tgui/interfaces/MafiaPanel.js [hunk 7]
```diff
@@ -131,7 +142,8 @@ export const MafiaPanel = (props, context) => {
             <Section
               title="Notes">
               <Table>
-                {!!roleinfo.action_log && roleinfo.action_log.map(log_line => (
+                {roleinfo !== undefined && !!roleinfo.action_log
+                && roleinfo.action_log.map(log_line => (
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/MafiaPanel.js:146]: All of this could be written as just `roleinfo?.action_log?.map(log_line => ...`

Comment (tralezab) [reply tgui/packages/tgui/interfaces/MafiaPanel.js:146]: 👍

**Comment by tralezab** [comment]
working on finishing map bugs (looking at you, necropolis and spider clan)

File: tgui/packages/tgui/interfaces/MafiaPanel.js [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+                && (<Fragment>Votes : {player.votes} </Fragment>)}
+                {
+                  !!player.actions && player.actions.map(action => {
+                    return (
+                      <Button
+                        key={action}
+                        onClick={
+                        // eslint-disable-next-line indent
+      () => act("mf_targ_action", { atype: action, target: player.ref })
+                        }>
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/MafiaPanel.js:95]: Please don't silence this obviously bad indent. Split act parameters over multiple lines.
```suggestion
                        onClick={() => act('mf_targ_action', {
                          atype: action,
                          target: player.ref,
                        })}>
```

File: tgui/packages/tgui/interfaces/MafiaPanel.js [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+              onClick={() => act("next_phase")}>Next Phase
+            </Button>
+            <Button
+              icon="home"
+              onClick={() => act("players_home")}>Send All Players Home
+            </Button>
+            <Button
+              icon="radiation"
+              onClick={() => act("new_game")}>New Game
+            </Button>
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/MafiaPanel.js:74]:
```suggestion
            <Button
              icon="arrow-right"
              onClick={() => act("next_phase")}>
              Next Phase
            </Button>
            <Button
              icon="home"
              onClick={() => act("players_home")}>
              Send All Players Home
            </Button>
            <Button
              icon="radiation"
              onClick={() => act("new_game")}>
              New Game
            </Button>
```

File: tgui/packages/tgui/interfaces/MafiaPanel.js [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+                icon="smile-beam"
+                color="good"
+                onClick={() => act("vote_innocent")}>INNOCENT!
+              </Button>
+              Use these buttons to vote the accused innocent or guilty!
+              <Button
+                icon="angry"
+                color="bad"
+                onClick={() => act("vote_guilty")}>GUILTY!
+              </Button>
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/MafiaPanel.js:116]:
```suggestion
              <Button
                icon="smile-beam"
                color="good"
                onClick={() => act("vote_innocent")}>
                INNOCENT!
              </Button>
              Use these buttons to vote the accused innocent or guilty!
              <Button
                icon="angry"
                color="bad"
                onClick={() => act("vote_guilty")}>
                GUILTY!
              </Button>
```

File: tgui/packages/tgui/interfaces/MafiaPanel.js [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+  const {
+    players,
+    actions,
+    phase,
+    roleinfo,
+    role_theme,
+    admin_controls,
+    judgement_phase,
+    timeleft,
+    all_roles } = data;
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/MafiaPanel.js:18]:
```suggestion
    all_roles,
   } = data;
```

File: tgui/packages/tgui/interfaces/MafiaPanel.js [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        <Flex>
+          {!!actions && actions.map(action => {
+            return (
+              <Flex.Item key={action}>
+                <Button
+                  onClick={() => act("mf_action", { atype: action })}>
+                  {action}
+                </Button>
+              </Flex.Item>);
+          })}
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/MafiaPanel.js:54]:
```suggestion
          {!!actions && actions.map(action => (
            <Flex.Item key={action}>
              <Button
                onClick={() => act("mf_action", { atype: action })}>
                {action}
              </Button>
            </Flex.Item>
          ))}
```

File: tgui/packages/tgui/interfaces/MafiaPanel.js [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+                    </Table.Cell>
+                  </Table.Row>
+                ))}
+              </Table>
+            </Section>
+          </Flex.Item>
+          <Flex.Item grow={2} basis={0}>
+            <Section
+              title="Notes">
+              <Table>
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/MafiaPanel.js:144]: You don't need a table here at all, just return plain boxes from the `roleinfo` loop.

File: tgui/packages/tgui/interfaces/MafiaPanel.js [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+                <Table.Cell bold>
+                  You are a {roleinfo.role}
+                </Table.Cell>
+              </Table.Row>
+              <Table.Row bold>
+                <Table.Cell>
+                  {roleinfo.desc}
+                </Table.Cell>
+              </Table.Row>
+            </Table>
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/MafiaPanel.js:42]: Same questionable use of tables, just use plain boxes.

File: _maps/map_files/Mafia/mafia_gothic.dmm [hunk 15]
```diff
@@ -0,0 +1,4 @@
+//MAP CONVERTED BY dmm2tgm.py THIS HEADER COMMENT PREVENTS RECONVERSION, DO NOT REMOVE
+"a" = (
+/turf/closed/indestructible/riveted,
+/area/mafia)
```
Comment (tralezab) [inline _maps/map_files/Mafia/mafia_gothic.dmm:4]: eob why

File: code/modules/mafia/controller.dm [hunk 16]
```diff
@@ -0,0 +1,578 @@
...
+			if(istype(possible_spawn, /obj/effect/landmark/mafia/town_center))
+				town_center_landmark = possible_spawn
+			else
+				landmarks += possible_spawn
+
+	current_setup_text = list()
+	for(var/rtype in setup_list)
+		for(var/i in 1 to setup_list[rtype])
+			all_roles += new rtype(src)
+			if(i == 1)
```
Comment (AnturK) [inline code/modules/mafia/controller.dm:77]: Just move this to the outer loop.

File: tgui/packages/tgui/interfaces/MafiaPanel.js [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+            <Button
+              icon="radiation"
+              onClick={() => act("new_game")}>
+              New Game
+            </Button>
+          </Section>
+        )}
+        <Section title="Players">
+          <LabeledList>
+            {!!players && players.map(player => { return (
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/MafiaPanel.js:82]: Use a shorter form:
```js
{!!players && players.map(player => (
```
Adjust parentheses at the end accordingly.

File: tgui/packages/tgui/interfaces/MafiaPanel.js [hunk 18]
```diff
@@ -0,0 +1,166 @@
+import { useBackend } from '../backend';
+import { Flex, Button, LabeledList, Section, Box, Table, TimeDisplay } from '../components';
+import { Fragment } from 'inferno';
+import { Window } from '../layouts';
+import { FlexItem } from '../components/Flex';
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/MafiaPanel.js:5]: Unused/redundant import.

**Comment by stylemistake** [comment]
@tralezab sorry in advance for bombarding your pull request with this feedback, not exactly your fault, but would be nice if you addressed it. 😄

File: code/modules/mafia/controller.dm [hunk 19]
```diff
@@ -0,0 +1,578 @@
...
+	if(message)
+		send_message(message)
+	for(var/datum/mafia_role/R in all_roles)
+		R.reveal_role(src)
+	phase = MAFIA_PHASE_VICTORY_LAP
+	next_phase_timer = addtimer(CALLBACK(src,.proc/end_game),victory_lap_period,TIMER_STOPPABLE)
+
+/datum/mafia_controller/proc/end_game()
+
+	map_deleter.generate() //remove the map, it will be loaded at the start of the next one
```
Comment (AnturK) [inline code/modules/mafia/controller.dm:225]: Using map generator here is bit pointless, just retrieve map template turfs and call empty on them

File: code/modules/mafia/controller.dm [hunk 20]
```diff
@@ -0,0 +1,578 @@
...
+	for(var/datum/mafia_role/role in all_roles)
+		var/mob/living/carbon/human/H = new(get_turf(role.assigned_landmark))
+		H.equipOutfit(player_outfit)
+		H.status_flags |= GODMODE
+		RegisterSignal(H,COMSIG_ATOM_UPDATE_OVERLAYS,.proc/display_votes)
+		var/datum/action/innate/mafia_panel/mafia_panel = new(null,src)
+		mafia_panel.Grant(H)
+		var/client/player_client = GLOB.directory[role.player_key]
+		if(player_client)
+			player_client.prefs.copy_to(H, antagonist = TRUE) ///i mean, it will cause github issue reports if this happens so sure throw it in
```
Comment (AnturK) [inline code/modules/mafia/controller.dm:335]: What ?

File: code/modules/mafia/controller.dm [hunk 21]
```diff
@@ -0,0 +1,578 @@
...
+					player.body.forceMove(get_turf(player.assigned_landmark))
+				if(failed.len)
+					to_chat(usr, "howdy armhulen :) here's the list of players that had null bodies on the pull:")
+					for(var/i in failed)
+						var/datum/mafia_role/fail = i
+						to_chat(usr, fail.player_key)
+	switch(action)
+		if("mf_lookup") //atype == Psychologist x1
+			var/raw_role = params["atype"]
+			var/role_name = copytext(raw_role, 1, length(raw_role) - 2) //cut the "x1" off the end
```
Comment (AnturK) [inline code/modules/mafia/controller.dm:426]: This is very ugly, pass the actual data you want.

File: code/modules/mafia/controller.dm [hunk 22]
```diff
@@ -0,0 +1,578 @@
...
+	var/datum/mafia_controller/parent
+
+/datum/action/innate/mafia_panel/New(Target,mf)
+	. = ..()
+	parent = mf
+
+/datum/action/innate/mafia_panel/Activate()
+	parent.ui_interact(owner)
+
+/datum/saymode/mafia
```
Comment (AnturK) [inline code/modules/mafia/controller.dm:561]: You should move this to where other saymodes are.

File: code/modules/mafia/map_pieces.dm [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+	requires_power = FALSE
+	has_gravity = STANDARD_GRAVITY
+	flags_1 = NONE
+
+/datum/map_template/mafia
+	var/description = ""
+	var/suffix = ""
+
+/datum/map_template/mafia/New()
+	mappath = "_maps/map_files/Mafia/" + suffix
```
Comment (AnturK) [inline code/modules/mafia/map_pieces.dm:39]: If it's static why do you do this, just write out whole path.

**Review by AnturK — CHANGES_REQUESTED** [review]
I'll review role part later.

**Comment by tralezab** [comment]
Resolving as I fix the issues in the reviews locally!

File: _maps/map_files/Mafia/mafia_gothic.dmm (hunk 15 above)
[the remark this answers is not in the document]
Comment (tralezab) [reply _maps/map_files/Mafia/mafia_gothic.dmm:4]: yeah, EOB. I don't >:(

File: code/modules/mafia/roles.dm [hunk 24]
```diff
@@ -0,0 +1,486 @@
...
+		to_chat(body,"<span class='warning'>You will block [target.body.real_name] tonight.</span>")
+
+/datum/mafia_role/lawyer/proc/try_to_roleblock(datum/mafia_controller/game)
+	if(current_target)
+		RegisterSignal(current_target,COMSIG_MAFIA_CAN_PERFORM_ACTION, .proc/prevent_action)
+
+/datum/mafia_role/lawyer/proc/release(datum/mafia_controller/game)
+	. = ..()
+	if(current_target)
+		UnregisterSignal(current_target)
```
Comment (AnturK) [inline code/modules/mafia/roles.dm:266]: No signal type.

File: code/modules/mafia/roles.dm [hunk 25]
```diff
@@ -0,0 +1,486 @@
...
+	desc = "You can visit a single member of the town to reveal their true role in the morning!"
+	revealed_outfit = /datum/outfit/mafia/psychologist
+
+	targeted_actions = list("Reveal")
+	var/datum/mafia_role/current_target
+	var/can_use = TRUE
+
+/datum/mafia_role/psychologist/New(datum/mafia_controller/game)
+	. = ..()
+	RegisterSignal(game,COMSIG_MAFIA_NIGHT_START,.proc/try_to_visit)
```
Comment (AnturK) [inline code/modules/mafia/roles.dm:284]: Why at night start. If they get killed it will still go through.

**Review by AnturK — CHANGES_REQUESTED** [review]

File: code/modules/mafia/roles.dm (hunk 25 above)
Comment (tralezab) [reply code/modules/mafia/roles.dm:284]: great point

**Comment by actioninja** [comment]
bundle conflict

**Pull request merged**